# Quick Introduction to Rust

Rust is described as having a steep learning curve.

The following topics will be covered here:

* Reviewing data types and variables in Rust
* Controlling variable ownership
* Building structs
* Metaprogramming with macros

## Why Rust?

With programming, there is usually a trade-off between speed/resources and development speed/safety.
Low-level languages such as C and C++ can give the developer fine grained control over the computer
with fast code execution and minimal resource consumption. However, this is not free. Manual memory
management can induce bugs and security vulnerabilities. On top of this, it takes more time and code
to solve a problem in a low-level language. As a result of this, C++ web frameworks do not take up
a large share of web development. Instead, it made sense to go for high-level programming languages
where developers can solve problems safely and quickly.

However, it has to be noted that this memory safety comes at a cost. Languages such as Python,
JavaScript, PHP, and Java keep track of all variables defined and their address, the data in that
memory address. When there are no more variables pointing to a memory address, the data in that
memory address gets deleted. This process is called garbage collection and consumes extra resources
and time.

With Rust, memory safety is ensured without the costly garbage collection process. Instead, the
compiler maps the variable, enforcing rules to ensure safety via a mechanism called the borrow
checker. Because of this, Rust has enabled rapid, safe problem solving with truly performant code,
thus breaking the speed/safety trade-off. As more data processing, traffic, and complex tasks are
lifted into the web stack, Rust, with its growing number of web frameworks and libraries, has now
become a viable choice for web development.

## Using Strings in Rust

```rust
fn print(input_string: String) {
    println!("{}", input_string);
}

fn main() {
    let test_string = String::from("Hello, Rob!");
    print(test_string);
}
```

The above code compiles and runs. Here we define a string using the `from` function in the `String`
object, and then passed it through our own `print` function to print it using Rust's built-in
`println!` function. (Technically, this is a macro; ! denotes that we can put multiple parameters
inside the parentheses).

Notice that the `print` function expects the `String` object to be passed through. If we try to
pass a literal string to that function it will throw the error:
`Expected 'String' found '&str' arguments to this function are in correct.`

To understand this we need to have a high-level understanding of stack and heap memory.

Stack memory is fast, static, and allocated at compile time. Heap memory is slower and allocated
at runtime. String literals can vary in side as they are the strings that we refer to, i.e.,
string literals are the actual strings, e.g. `A text` or `Another text` so their size will change.
String objects, on the other hand, have a fixed size in the stack that consists of a reference to
the string literal in the heap. When we pass a string literal to our own `print` function, it will
have no idea of the size of the string literal being passed through. String literals can be
converted into strings with `to_string`:

```rust
fn print(input_string: String) {
	println!("{}", input_string);
}

fn main() {
	print("Hello, Rob!".to_string());
}
```

In the example above, we converted the string literal into a string object just before passing it
through the print function.

We can also get the print function to accept a string literal reference by borrowing it using the
`&` operator:

```rust
fn print(input_string: &str) {
    println!("{}", input_string);
}

fn main() {
    print("Hello, Rob");
}
```

To get the string literal from an string object use the `as_str` method.

```rust
fn print(input_string: &str) {
    println!("{}", input_string);
}

fn main() {
    print(String::from(":)").as_str());
}
```

## Understanding Integers and Floats

Rust has signed integers (denoted by i) and unsigned integers (denoted by u) that consist of
8, 16, 32, 64, and 128 bits. We need to understand the range of numbers allowed in terms of bits.
Because binary is either 0 or 1, we can calculate the integer range by raising 2 to the power of
the number of bits. For example, for 8 bits, 2 to the power of 8 equates to 256. Considering the
0, this means that an u8 integer should have a range of 0 to 255:

```rust
let number: u8 = 255;
```

```rust
let number: i8 = 127;
```

In terms of floats, Rust accommodates f32 and f64 floating points, which can be both negative and
positive. Declaring a floating-point variable requires the same syntax as integers:

```rust
let float: f32 = 20.6;
```

It is also possible to annotate numbers with suffixes:

```rust
let x - 1u8; // x has the value of an 8-bit unsigned integer 1,
```

## Storing data in vectors and arrays

Rust stores sequenced data in vectors and arrays. Arrays are generally immutable and don't have push
functions. They also only accommodate one data type. This can be managed using structs and traits.
You can define and loop through arrays with fairly standard syntax:

```rust
fn main() {
    let int_array: [i32; 3] = [1, 2, 3];

    for i in int_array.iter() {
        println!("{}", i);
    }

    let str_vector: Vec<&str> = vec!["one", "two", "three"];

    for i in str_vector.iter() {
        println!("{}", i);
    }

    println!("{}", str_vector[1]);
}
```

By default variables defined in Rust are not mutable. To change that we need to put the `mut` keyword
in front of the variable's name:

```rust
let mut str_vector: Vec<&str> = vec!["one", "two", "three"];
```

This also works for strings and numbers. While it might be tempting to define everything as `mut`, this
forced immutability not only has performance benefits, but it also improves the safety. If you are not
expecting a variable to change in a complex system, then not allowing it to mutate will throw up the
error right then as opposed to allowing silent bugs to run in your system.

## Mapping data with hash maps

In some languages, hash maps are referred to as dictionaries. In order to define a hash map in Rust, we
must import the hash maps from the standard library. Once we've defined a new hash map, we can insert
and entry, get it out of the hash map, and then print it:

```rust
use std::collections::HashMap;

fn main() {
    let mut general_map: HashMap<&str, i8> = HashMap::new();
    general_map.insert("test", 25);
    let outcome: Option<&i8> = general_map.get("test");
    println!("{}", outcome.unwrap());
}
```

The `get` method from the HashMap does not return the type that was used in the declaration, instead it
returns an `Option` enum instead. This is because the `get` method could fail. We could pass it a key
that does not exist. Therefore we have to unwrap the option to get the value that we're aiming to get.
However, directly unwrapping the result can result in an error being raised. Because `Optional` is
either `Some` or `None`, we can explore Rust's `match` statement to handle the outcome:

```rust
use std::collections::HashMap;

fn main() {
    let mut general_map: HashMap<&str, i8> = HashMap::new();
    general_map.insert("test", 25);
    match general_map.get("tist") {
        None => println!("It failed."),
        Some(result) => println!("Here is the result: {}.", result)
    }
}
```

The arrows in the match statement can have their own code blocks.

Calling the `insert` function again with the same key will merely update the value under that key.
Calling the `remove` function from the hash map with the desired key will remove the entry if it
exists.

[HashMap Docs](https://doc.rust-lang.org/beta/std/collections/struct.HashMap.html).

## Handling Results and Errors

Like other languages, Rust throws and handles errors. It manages errors through two different types:
`Option` and `Result`. `Option` only returns `None` or `Some`. `Result` only returns `Err` or `Some`.

These are fairly similar, however, if `Err` is exposed, the `Rust` program panics and the program
crashes with what is the outcome of `Err`. While there will be plenty of opportunities to throw
errors, we will also want to throw our own when needed. When system becomes more complex, it can be
handy to purposefully throw errors if there is any undesired behavior. A good example is inserting
data into a Redis cache.

Technically, there is nothing stopping us from inserting a range of keys into Redis. In order to
prevent this, if the key is not an expected variant of what we want, we should throw an error.
Let's demonstrate how to throw an error, depending on the data:

```rust
fn error_check(check: bool) -> Result<i8, &'static str> {
    if check == true {
        Err("This is an error.")
    } else {
        Ok(1)
    }
}

fn main() {
    let result: i8 = error_check(true).unwrap();
    println!("{}", result);
}
```

Note that there is no `return` keyword. This is because the function returns the final expression
in the function when there is no semicolon at the end of the expression.

This `Result` wrapper gives us a lot of control of the outcome. Instead of throwing try and except
blocks, we can wait until we're ready to handle the error. We can build a simple error handling
function with a `match` statement:

```rust
fn error_check(check: bool) -> Result<i8, &'static str> {
    if check == true {
        Err("This is an error.")
    } else {
        Ok(1)
    }
}

fn describe_result(result: Result<i8, &'static str>) {
    match result {
        Ok(x) => println!("it's a result of: {}", x),
        Err(x) => println!("{}", x)
    }
}

fn main() {
    let result: Result<i8, &'static str> = error_check(false);
    describe_result(result);
}
```

In the wild, this comes in useful when we must roll back a database entry or clean up a process
before throwing an error. We also have to note the typing for `Result`. In this result, we
return an i8 integer (we can return other variables), but we can also return a reference to a
string literal that has the `'static` notation. This is the lifetime notation. The `'static`
notation is telling the compiler that the error string will stay around for the entire runtime
of the program.

This makes sense as we would hate to lose the error message because we moved out of a scope.
Also, it's an error, so we should be ending the program soon. If we want to tolerate an outcome,
we should be reaching for the option and handling `None`. We can also signpost a little more
with the expect function as opposed to using `unwrap`. It still unwraps the result, but adds an
extra message in the error trace.

```rust
let result: i8 = error_check(true).expect("This has been caught");
```

We can also directly throw errors with the `panic` function:

```rust
panic!("Throwing some error");
```

We can also check for an error using the `is_err` function:

```rust
result.is_err()
```

This returns a bool value. As we can see, Rust supports a range of error handling. It is advised
to keep these as simple as possible. For most processes in a simple web app, unwrapping straight
away and throwing the error as soon as possible will manage most situations.

## Controlling variable ownership

As Rust does not have a garbage collector, it maintains memory safety by enforcing strict rules
around variable ownership that are enforced when compiling. If these rules are understood early,
the helpful compiler makes it straightforward to adhere to them. Rust's compile-time checking is
done to protect against the following memory errors:

* Use after frees: This is where memory is used after it has been freed, which can cause crashes.
It can also allow hackers to execute code via this memory address.

* Dangling pointers: This is where a reference points to a memory address that no longer houses
the data that the pointer was referencing. Essentially, this pointer now points to null or random
data.

* Double frees: This is where allocated memory is freed, and then freed again. This can cause the
problem to crash and increases the risk of sensitive data being revealed. This also enables a
hacker to execute arbitrary code.

* Segmentation faults: This is where the program tries to access the memory it's not allowed  to
access.

* Buffer overrun: An example of this is reading off the end of an array. This can cause the
program to crash.

Protection is achieved by Rust following ownership rules. These ownership rules flag code that
can lead to the memory errors just mentioned. If these rules are broken they are flagged up as
compile-time errors. These are defined here:

* Values are owned by the variables assigned to them.
* As soon as a variable goes out of scope, it is deallocated from the memory it is occupying.
* Values can be used by other variables, as long as we adhere to the following rules:
  * Copy: This is where the value is copied. Once it has been copied, the new variable owns
	the value, and the existing variable also owns its own value.
  * Move: This is where the value is moved from one variable to another. However, unlike copy,
	the original variable no longer owns the value.
  * Immutable borrow: This is where another variable can reference the value of another variable.
	If the variable that is borrowing the value falls out of scope, the value is not deallocated
	from memory as the variable borrowing the value does not have ownership.
  * Mutable borrow: This is where another variable can reference and write the value of another
	variable. If the variable that is borrowing the value falls out of scope, the value is not
	deallocated from memory as the variable borrowing the value does not have ownership.

Considering that scopes play a big role in ownership rules, let's explore them in more detail
next.

### Scopes

The key rule to remember when it comes to ownership in Rust is that when `let` is used to
create a variable, that variable is the only one that owns the resource. Therefore, if the
resource is moved or reassigned, then the initial variable no longer owns the resource.

Once the scope has ended, then the variable and the resource are deleted. A good way to
demonstrate this is through scopes. Scopes in Rust are defined by curly brackets. The classic
way of demonstrating this is through the following example:

```rust
fn print_number(number: String) {
    println!("{}", number);
}

fn main() {
    let one: String = String::from("one");
    print_number(one);
    println!("{}", one);
}
```

This code throws the error:

```bash
   Compiling playground v0.0.1 (/playground)
error[E0382]: borrow of moved value: `one`
 --> src/main.rs:8:20
  |
6 |     let one: String = String::from("one");
  |         --- move occurs because `one` has type `String`, which does not implement the `Copy` trait
7 |     print_number(one);
  |                  --- value moved here
8 |     println!("{}", one);
  |                    ^^^ value borrowed here after move
  |
note: consider changing this parameter type in function `print_number` to borrow instead if owning the value isn't necessary
 --> src/main.rs:1:25
  |
1 | fn print_number(number: String) {
  |    ------------         ^^^^^^ this parameter takes ownership of the value
  |    |
  |    in this function
  = note: this error originates in the macro `$crate::format_args_nl` which comes from the expansion of the macro `println` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider cloning the value if the performance cost is acceptable
  |
7 |     print_number(one.clone());
  |                     ++++++++

For more information about this error, try `rustc --explain E0382`.
error: could not compile `playground` due to previous error
```

The stem of the error has occurred because String does not implement a copy trait. This is not
surprising as we know that String is a type of wrapper implemented as a vector of bytes. This
vector holds a reference to str, the capacity of str in the heap memory, and the length of str:

|•|6|3| = String in the Stack memory
 |
 |
|O|n|e| | | | = str in th heap memory

Having multiple references to the value breaks our rules. Passing one through our print function
moves it into another scope, which is then destroyed. If we passed ownership to a function but
still allowed references outside the function later on, these references will be pointing to freed
memory which is unsafe.

The compiler is very helpful in telling us that the variable has been moved, which is why it cannot
print it. There is another hint. You can see that the built-in print method tries to borrow String.
When you borrow a variable, you can access the data, but only for as long as you need it. Borrowing
can be done by using the & operator. Therefore, we can get around the issue in the code above with
the following code:

```rust
fn alter_number(number: &mut String) {
    number.push("!".chars().next().unwrap());
}

fn print_number(number: &String) {
    println!("{}", number);
}

fn main() {
    let mut one: String = String::from("one");
    print_number(&one);
    alter_number(&mut one);
    println!("{}", one);
}
```

In the preceding code, we borrowed the string to print it. In the second function, we did a mutable
borrow, meaning that we can alter the value. We then defined a string literal, converted it into an
array of chars, called the next function since it is a generator, and then unwrapped it and appended
it to the string. We can see by the final print statement that the one variable has been changed.

If we were to try and change the value in the print_number function, we would get an error because
it's not a mutable borrow, despite one being mutable. When it comes to immutable borrows, we can make
as many as we like. For instance, if we are borrowing for a function, the function does not need to
own the variable. If there is a mutable borrow, then only one mutable borrow can exist at one time,
and during that lifetime, no immutable borrows can be made. This is to avoid data races.

With integers this is easier as they implement the copy trait. This means that we don't have to borrow
when passing the copy trait into a function. It's copied for us.

```rust
fn alter_number(number: &mut i8) {
    *number += 1
}

fn print_number(number: i8) {
    println!("{}", number);
}

fn main() {
    let mut one: i8 = 1;
    print_number(one);
    alter_number(&mut one);
    println!("{}", one);
}
```

Here we can see that the integer isn't moved into print_number; it's copied. However, we still have
to pass a mutable reference if we want to alter the number. We can also see that we've added a `*`
operator to the number when altering it. This is a dereference. By performing this, we have access
to the integer value that we're referencing. Remember that we can directly pass the integer into
the print_number function because we know the maximum size of all i8 integers.

## Running through lifetimes

Remember that borrow is not sole ownership. Because of this, there is a risk that we could reference a
variable that's deleted. Consider the following:

```rust
fn main() {
    let one;
    {
        let two: i8 = 2;
        one = &two;
    } // ------------------> 'two' lifetime stops here
    println!("r: {}", one);
}
```

This gives the following error:

```bash
   Compiling playground v0.0.1 (/playground)
error[E0597]: `two` does not live long enough
 --> src/main.rs:5:15
  |
5 |         one = &two;
  |               ^^^^ borrowed value does not live long enough
6 |     } // ------------------> 'two' lifetime stops here
  |     - `two` dropped here while still borrowed
7 |     println!("r: {}", one);
  |                       --- borrow later used here

For more information about this error, try `rustc --explain E0597`.
error: could not compile `playground` due to previous error
```

Since the reference is defined in the inner scope, it's deleted at the end of the inner scope, meaning that
the end of its lifetime is at the end of the inner scope. However, the lifetime of that one variable, carries
on to the end of the scope of the main function. Therefore, the lifetimes are not equal.

While it is great that this is flagged when compiling, Rust does not stop here. This concept also translates
into functions. Let's say that we build a function that references two integers, compares them, and returns
the highest integer reference. The function is an isolated piece of code. In this function, we can denote the
lifetime of the two integer. This is done by using the `'` prefix, which is a lifetime notation. The names of
the notations can be anything you wish, but it's a general convention to use `a,b,c` and so on. Let's look at
an example:

```rust
fn get_highest<'a>(first_number: &'a i8, second_number: &'a i8) -> &'a i8 {
    if first_number > second_number {
        first_number
    } else {
        second_number
    }
}

fn main() {
    let one: i8 = 1;
    {
        let  two: i8 = 2;
        let outcome: &i8 = get_highest(&one, &two);
        println!("{}", outcome);
    }
}
```

As we can see, the first and second lifetimes have the same notation of `a`. They will both have to be present
for the duration of the function. We also have to note that the function returns an i8 integer with the
lifetime of a. Therefore, the compiler knows that we cannot rely on the outcome outside the inner loop.
However, we may want to just use the two variable that is defined in the inner scope for reference in the
function, but not for the result.

This might be a little convoluted, so to demonstrate this, let's develop a function that checks the `one`
variable. If `one` is lower than `two`, then we return zero; otherwise, we return the value of `one`:

```rust
fn filter<'a, 'b>(first_number: &'a i8, second_number: &'b i8) -> &'a i8 {
    if first_number < second_number {
        &0
    } else {
        first_number
    }
}

fn main() {
    let one: i8 = 1;
    let outcome: &i8;
    {
        let two: i8 = 2;
        outcome = filter(&one, &two);
    }
    println!("{}", outcome);
}
```

Here, we assigned the lifetime of 'a to first_number, and the lifetime of 'b to second_number. Using 'a
and 'b, we are telling the compiler that the lifetimes are different. We then tell the compiler in the
return typing of the function that the function returns an i8 integer with the lifetime of 'a. Therefore,
we can rely on the result of the filter function, even if the lifetime of second_number finishes.

If we switch the second_number lifetime of 'a, we get the following expected error:

```bash
   Compiling playground v0.0.1 (/playground)
error[E0597]: `two` does not live long enough
  --> src/main.rs:14:32
   |
14 |         outcome = filter(&one, &two);
   |                                ^^^^ borrowed value does not live long enough
15 |     }
   |     - `two` dropped here while still borrowed
16 |     println!("{}", outcome);
   |                    ------- borrow later used here

For more information about this error, try `rustc --explain E0597`.
error: could not compile `playground` due to previous error
```

This happens because since you are specifying the lifetime of `second_number` also as `'a`, the compiler
will take the lifetime of that argument (`two`) to use as the lifetime of `'a`, so that ends in the inner
scope.

Even though we're still just returning `first_number` that is available in the outer scope, we're telling
the compiler that we're returning a variable with the `'a` lifetime, which is assigned to `first_number`
and `second_number`. The compiler is going to side with the shortest lifetime to be safe when both lifetimes
are denoted to be the same in the functions.

Now that we understand the quicks behind data types, borrowing, and lifetimes, we're ready to build our
structs that have the functionality to create a hash map that accepts a range of data type.

## Building Structs

In dynamic languages, classes have been the bedrock of developing data structures with custom
functionality. In Rust, structs are used to define data structures with functionality. To mimic
a class, we can define a Human struct:

```rust
struct Human {
	name: String,
	age: i8,
	current_thought: String
}

impl Human {
	fn new(input_name: &str, input_age: i8) -> Human {
		return Human {
			name: input_name.to_string(),
			age: input_age,
			current_thought: String::from("nothing")
		}
	}
	fn with_thought(mut self, thought: &str) -> Human {
		self.current_thought = thought.to_string();
		return self
	}

	fn speak(&self) -> () {
		println!("Hello, my name is {} and I'm {} years old.", &self.name, &self.age);
	}
}

fn main() {
	let developer = Human::new("Robson William", 30);
	developer.speak();
	println!("Currently I'm thinking {}", developer.current_thought);

	let new_developer = Human::new("Digory", 30).with_thought("I'm hungry");
	new_developer.speak();
	println!("Currently I'm thinking {}", new_developer.current_thought);
}
```

Here, we will exploit enums to allow the hash map to accept an integer or a string:

```rust
use std::collections::HashMap;

enum AllowedData {
	S(String),
	I(i8)
}

struct CustomMap {
	body: HashMap<String, AllowedData>
}

impl CustomMap {

	fn new() -> CustomMap {
		return CustomMap{body: HashMap::new()}
	}

	fn get(&self, key: &str) -> &AllowedData {
		return self.body.get(key).unwrap()
	}

	fn insert(&mut self, key: &str, value: AllowedData) -> () {
		self.body.insert(key.to_string(), value);
	}

	fn display(&self, key: &str) -> () {
		match self.get(key) {
			AllowedData::I(value) => println!("{}", value),
			AllowedData::S(value) => println!("{}", value)
		}
	}
}

fn main() {
	let mut map = CustomMap::new();
	map.insert("test", AllowedData::I(8));
	map.insert("testing", AllowedData::S("test value".to_string()));

	map.display("test");
	map.display("testing");
}
```

## Verifying with traits

Enums can empower our structs so that they can handle multiple types. This can also be translated for any
type of function or data structure. However, this can lead to a lot of repetition. Traits can help adding
the same functionality to different data types.

```rust
trait CanEdit {
	fn edit(&self) {
		println!("user is editing");
	}
}

trait CanCreate {
	fn create(&self) {
		println!("user is creating");
	}

trait CanDelete {
	fn delete(&self) {
		println!("user is deleting");
	}
}
```

If a struct implements a trait, then it can use and overwrite the functions defined in the trait block.

```rust
struct AdminUser {
	name: String,
	password: String
}

impl CanDelete for AdminUser {}
impl CanCreate for AdminUser {}
impl CanEdit for AdminUser {}
```

Now that our user struct has implemented all three traits, we can create a function that only allows users
inside that have the CanDelete trait implemented:

```rust
fn delete<T: CanDelete>(user: T) -> () {
	user.delete();
}
```

Similar to the lifetime notation, we use angle brackets before the input definitions to define T as a
CanDelete trait. If we create a general user struct and we don't implement the CanDelete trait for it,
Rust will fail to compile if we try to pass the general user through the delete function; it will
complain, stating that it does not implement the CanDelete trait.

The following code implements a user struct that inherits from a base user struct and has traits that can
allow us to use the user struct in different functions. Rust does not directly support inheritance. However,
we can combine structs with basic composition:

```rust
trait CanEdit {
	fn edit(&self) {
		println!("user is editing");
	}
}

trait CanCreate {
	fn create(&self) {
		println!("user is creating");
	}
}

trait CanDelete {
	fn delete(&self) {
		println!("user is deleting");
	}
}

struct AdminUser {
	name: String,
	password: String,
}

impl CanDelete for AdminUser {}
impl CanCreate for AdminUser {}
impl CanEdit for AdminUser {}

fn delete<T: CanDelete>(user: T) -> () {
	user.delete();
}

struct BaseUser {
	name: String,
	password: String
}

struct GeneralUser {
	super_struct: BaseUser,
	team: String
}

impl GeneralUser {

	fn new(name: String, password: String, team: String) -> GeneralUser {
		return GeneralUser{super_struct: BaseUser{name, password}, team: team}
	}
}

impl CanEdit for GeneralUser {}

impl CanCreate for GeneralUser {
	fn create(&self) -> () {
		println!("{} is creating under a {} team", self.super_struct.name, self.team);
	}
}

fn main() {
    let u = GeneralUser::new("Bob".to_string(), "1".to_string(), "Alpha".to_string());
    delete(u);
}
```

The code in the `main` function above threw this error:

```bash
   Compiling playground v0.0.1 (/playground)
error[E0277]: the trait bound `GeneralUser: CanDelete` is not satisfied
  --> src/main.rs:59:12
   |
59 |     delete(u);
   |     ------ ^ the trait `CanDelete` is not implemented for `GeneralUser`
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `CanDelete` is implemented for `AdminUser`
note: required by a bound in `delete`
  --> src/main.rs:28:14
   |
28 | fn delete<T: CanDelete>(user: T) -> () {
   |              ^^^^^^^^^ required by this bound in `delete`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `playground` due to previous error
```
